sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveUnion.scala

```diff
@@ -20,137 +20,66 @@ package org.apache.spark.sql.catalyst.analysis
-    case u: UpdateFields if u.resolved =>
-      u.evalExpr match {
-        case i @ If(IsNull(_), _, CreateNamedStruct(fieldExprs)) =>
-          val sorted = sortFieldExprs(fieldExprs)
-          val newStruct = CreateNamedStruct(sorted)
-          i.copy(trueValue = Literal(null, newStruct.dataType), falseValue = newStruct)
-        case CreateNamedStruct(fieldExprs) =>
-          val sorted = sortFieldExprs(fieldExprs)
-          val newStruct = CreateNamedStruct(sorted)
-          newStruct
-        case other =>
-          throw new IllegalStateException(s"`UpdateFields` has incorrect expression: $other. " +
-            "Please file a bug report with this error message, stack trace, and the query.")
-      }
-  }
-
-  /**
-   * Adds missing fields recursively into given `col` expression, based on the target `StructType`.
-   * This is called by `compareAndAddFields` when we find two struct columns with same name but
-   * different nested fields. This method will find out the missing nested fields from `col` to
-   * `target` struct and add these missing nested fields. Currently we don't support finding out
-   * missing nested fields of struct nested in array or struct nested in map.
-   */
-  private def addFields(col: NamedExpression, target: StructType): Expression = {
+  private def addFields(col: Expression, targetType: StructType): Expression = {
     assert(col.dataType.isInstanceOf[StructType], "Only support StructType.")
 
     val resolver = conf.resolver
-    val missingFieldsOpt =
-      StructType.findMissingFields(col.dataType.asInstanceOf[StructType], target, resolver)
-
-    // We need to sort columns in result, because we might add another column in other side.
-    // E.g., we want to union two structs "a int, b long" and "a int, c string".
-    // If we don't sort, we will have "a int, b long, c string" and
-    // "a int, c string, b long", which are not compatible.
-    if (missingFieldsOpt.isEmpty) {
-      sortStructFields(col)
-    } else {
-      missingFieldsOpt.map { s =>
-        val struct = addFieldsInto(col, s.fields)
-        // Combines `WithFields`s to reduce expression tree.
-        val reducedStruct = struct.transformUp(OptimizeUpdateFields.optimizeUpdateFields)
-        val sorted = sortStructFieldsInWithFields(reducedStruct)
-        sorted
-      }.get
+    val colType = col.dataType.asInstanceOf[StructType]
+
+    val newStructFields = mutable.ArrayBuffer.empty[Expression]
+
+    val targetStructFields = targetType.fields.foreach { expectedField =>
+      val currentField = colType.fields.find(f => resolver(f.name, expectedField.name))
+
+      val newExpression = (currentField, expectedField.dataType) match {
+        case (Some(cf), expectedType: StructType) if cf.dataType.isInstanceOf[StructType] =>
+            val extractedValue = ExtractValue(col, Literal(cf.name), resolver)
+            val combinedStruct = addFields(extractedValue, expectedType)
+            if (extractedValue.nullable) {
```

Comment:
It's hard to see why null handling is needed here. I think we should move the null handling to where we return `CreateNamedStruct`

Reply:
which is https://github.com/apache/spark/pull/33040/files#diff-84dd17265dcadd59f6ad9e649203d38b808485c7b5bd3937136222378f2ed27dR79

Reply:
Yeah I just copied this from the old version. @viirya is there a reason all these nullable checks needed to be added? I don't know what you mean by move the null handling to where we return `CreateNamedStruct` though

Reply:
Err maybe I didn't copy that directly from somewhere, just saw it used other places. Not sure if/where/when we need the nullable checks

Reply:
Do you mean:

```scala
val newExpr = CreateNamedStruct(existingExprs)
If (expr.nullable) {
  If(IsNull(expr), Literal(null, newExpr.dataType), newExpr)
} else {
  newExpr
}
```

Otherwise I don't find nullable check.

Reply:
Yea, the code posted by @viirya is more widely used in the spark codebase. The rationale is pretty simple: CreateNamedStruct never return null and we need to handle the case if input is null.

Reply:
Yeah, why was that needed? I copied that I guess because I thought it might be needed here as well for some reason. Is that basically trying to keep nulls down to the lowest level of the struct, instead of taking a null struct and making a non-null struct with all null values?

Reply:
Moved the null check to the CreatedNamedStruct